Userland/xargs.cpp

```diff
@@ -0,0 +1,307 @@
+                return Unget;
+            } else {
+                child_argv.append(strndup(item.characters_without_null_termination(), item.length()));
+                ++items_used_for_this_command;
+            }
+        }
+
+        return Continue;
+    });
+
+    if (!fail && !child_argv.is_empty())
+        fail = !run_command(child_argv, verbose, is_stdin, devnull_fd);
+
+    if (!is_stdin)
+        fclose(fp);
+
+    return fail ? 1 : 0;
+}
+
+bool read_items(FILE* fp, char entry_separator, Function<Decision(StringView)> callback)
+{
+    bool fail = false;
+
+    for (;;) {
+        char* item = nullptr;
+        size_t buffer_size = 0;
+
+        auto item_size = getdelim(&item, &buffer_size, entry_separator, fp);
+
+        if (item_size < 0) {
+            // getdelim() will return -1 and set errno to 0 on EOF.
+            if (errno != 0) {
+                perror("getdelim");
+                fail = true;
+            }
+            break;
+        }
+
+        Decision decision;
+        do {
+            decision = callback(item);
+            if (decision == Stop) {
+                free(item);
+                return true;
+            }
+        } while (decision == Unget);
+
+        free(item);
+    }
+
+    return fail;
+}
+
+bool run_command(Vector<char*>& child_argv, bool verbose, bool is_stdin, int devnull_fd)
+{
+    child_argv.append(nullptr);
+
+    if (verbose) {
+        StringBuilder builder;
+        builder.join(" ", child_argv);
+        fprintf(stderr, "xargs: %s\n", builder.to_string().characters());
+        fflush(stderr);
+    }
+
+    auto pid = fork();
+    if (pid < 0) {
+        perror("fork");
+        return false;
+    }
+
+    if (pid == 0) {
+        if (is_stdin)
+            dup2(devnull_fd, STDIN_FILENO);
+
+        execvp(child_argv[0], child_argv.data());
+        exit(1);
+    }
+
+    for (auto* ptr : child_argv)
+        free(ptr);
+
+    child_argv.clear();
+
+    int wstatus = 0;
+    if (waitpid(pid, &wstatus, 0) < 0) {
+        perror("waitpid");
+        return false;
+    }
+
+    if (WIFEXITED(wstatus)) {
+        if (WEXITSTATUS(wstatus) != 0)
+            return false;
+    } else {
+        return false;
+    }
+
+    return true;
+}
+
+ParsedInitialArguments::ParsedInitialArguments(Vector<const char*>& arguments, const StringView& placeholder)
+{
+    m_all_parts.ensure_capacity(arguments.size());
+    bool some_argument_has_replacement = false;
+
+    for (auto argument : arguments) {
+        StringView arg { argument };
+        Vector<StringView> parts;
+        size_t previous_index = 0;
+        if (!placeholder.is_empty()) {
+            // Split the argument into segments delimited by the placeholder.
+            if (auto first_index = arg.find_first_of(placeholder); first_index.has_value()) {
+                some_argument_has_replacement = true;
+                while (first_index.has_value()) {
+                    auto current_index = first_index.value();
```

Comment:
So `current_index` is simply the non-optional `first_index`? Perhaps the naming could be improved, though I don't have immediate suggestions.

Reply:
Eh, `first_index` is named badly, it'd be `maybe_placeholder_start_index`
and `current_index`...`placeholder_start_index`, I guess?

Reply:
Yeah, that sounds better.

I also wish C++ would let you rebind names, like

```c++
auto foo_bar = baz();
if (foo_bar.is_error())
    return foo_bar.error();
auto foo_bar = foo_bar.value();
```

(Or even
```c++
auto foo_bar = baz()?;
```
but that's going too far with the idea.)